# Задача на 6-ом шаге урока

**⚖️ Взвешенный блендинг**

Задание: У тебя есть информация о точности модели на лидерборде (accuracy_score). Рассчитай веса, с которыми нужно будет взвесить предсказания моделей, чтобы они были пропорциональны accuracy_score, а их сумма равнялась единице.

Примечание: Ввод и вывод уже написан за вас.

In [230]:
# Формат ввода:

# В первой строке идет число n <= 100 — количество моделей, которые участвуют в блендинге. 
# Во второй строке через пробел записаны n чисел — значения точности каждой из этих моделей на лидерборде (от 0 до 1).
# Формат вывода:

# В единственной строке выведите n чисел — веса, с которыми нужно взвесить соответствующие модели. 
# Значения весов округлите до 7 знака после запятой. Исходный порядок моделей важен. 

import numpy as np

n = 3
s = '0.3 0.5 0.4'.split()
scores = []
for i in range(n):
    scores.append(float(s[i]))

## Your code here ...
weights = np.round((np.array(scores) / np.sum(scores)), 7)
weights

array([0.25     , 0.4166667, 0.3333333])

Sample Input:

3

0.3 0.5 0.4

Sample Output:

0.25 0.4166667 0.3333333

# Задача на 7-ом шаге урока

**🏄‍♂️ Hard voting**

Датасет с прогнозами моделей тут:

In [231]:
import pandas as pd

In [232]:
path = '../data/make_hard_blend.csv'

df = pd.read_csv(path)
df.head()

,car_id,target_class,model_1,model_2,model_3,model_4,model_5
0,car_0,NaN,electro_bug,gear_stick,electro_bug,electro_bug,engine_ignition
1,car_1,NaN,engine_check,engine_overheat,electro_bug,engine_overheat,engine_overheat
2,car_2,NaN,another_bug,electro_bug,engine_overheat,engine_overheat,gear_stick
3,car_3,NaN,electro_bug,electro_bug,electro_bug,wheel_shake,electro_bug
4,car_4,NaN,engine_ignition,another_bug,engine_check,engine_check,engine_check


In [233]:
df

,car_id,target_class,model_1,model_2,model_3,model_4,model_5
0,car_0,NaN,electro_bug,gear_stick,electro_bug,electro_bug,engine_ignition
1,car_1,NaN,engine_check,engine_overheat,electro_bug,engine_overheat,engine_overheat
2,car_2,NaN,another_bug,electro_bug,engine_overheat,engine_overheat,gear_stick
3,car_3,NaN,electro_bug,electro_bug,electro_bug,wheel_shake,electro_bug
4,car_4,NaN,engine_ignition,another_bug,engine_check,engine_check,engine_check
...,...,...,...,...,...,...,...
1908,car_1908,NaN,break_bug,break_bug,engine_overheat,break_bug,another_bug
1909,car_1909,NaN,break_bug,break_bug,wheel_shake,break_bug,gear_stick
1910,car_1910,NaN,engine_overheat,engine_overheat,engine_overheat,engine_overheat,break_bug
1911,car_1911,NaN,engine_check,engine_check,engine_check,electro_bug,break_bug


In [234]:
df['target_class'] = df.apply(lambda x: x[['model_1', 'model_2', 'model_3', 'model_4', 'model_5']].mode()[0], axis=1)

Файл с решением должен выглядет так с заполненными значениями target_class

In [235]:
result = df[['car_id', 'target_class']]
result.head(10)

,car_id,target_class
0,car_0,electro_bug
1,car_1,engine_overheat
2,car_2,engine_overheat
3,car_3,electro_bug
4,car_4,engine_check
5,car_5,engine_fuel
6,car_6,engine_check
7,car_7,wheel_shake
8,car_8,break_bug
9,car_9,another_bug


In [236]:
result.to_csv('../subs/cars_hard_voting.csv', index=False)
pd.read_csv('../subs/cars_hard_voting.csv').head(10)

,car_id,target_class
0,car_0,electro_bug
1,car_1,engine_overheat
2,car_2,engine_overheat
3,car_3,electro_bug
4,car_4,engine_check
5,car_5,engine_fuel
6,car_6,engine_check
7,car_7,wheel_shake
8,car_8,break_bug
9,car_9,another_bug


In [237]:
path = '../data/sb_sample_submission.csv'

subs = pd.read_csv(path).head()
subs.head()

,car_id,target_class
0,car_0,NaN
1,car_1,NaN
2,car_2,NaN
3,car_3,NaN
4,car_4,NaN


# Задача на 8-ом шаге урока

**🤝Soft Voting**

Датасета с предсказаниями поломок машин от разных моделей:

In [238]:
cb_preds = pd.read_csv('../data/catboost_proba.csv')
cb_preds.head()

,another_bug_proba,break_bug_proba,electro_bug_proba,engine_check_proba,engine_fuel_proba,engine_ignition_proba,engine_overheat_proba,gear_stick_proba,wheel_shake_proba
0,0.112482,0.117905,0.102009,0.115111,0.105105,0.111646,0.108378,0.128776,0.098588
1,0.131855,0.114520,0.111594,0.113760,0.099113,0.108545,0.105714,0.119711,0.095189
2,0.112382,0.117429,0.104545,0.106867,0.105233,0.112577,0.112477,0.128264,0.100227
3,0.117932,0.117098,0.102200,0.119365,0.101529,0.106561,0.110749,0.130157,0.094410
4,0.109022,0.117847,0.103836,0.117278,0.103819,0.109251,0.115560,0.133949,0.089437


In [239]:
lgbm_preds  = pd.read_csv('../data/lgbm_proba.csv')
lgbm_preds .head()

,another_bug_proba,break_bug_proba,electro_bug_proba,engine_check_proba,engine_fuel_proba,engine_ignition_proba,engine_overheat_proba,gear_stick_proba,wheel_shake_proba
0,0.120016,0.113801,0.114904,0.108070,0.118519,0.105637,0.121196,0.113500,0.084358
1,0.159507,0.211517,0.084138,0.099253,0.082684,0.087642,0.113180,0.103527,0.058553
2,0.120716,0.111604,0.107702,0.111085,0.104361,0.126828,0.121338,0.118276,0.078090
3,0.111751,0.117286,0.105884,0.118102,0.108398,0.114491,0.109710,0.130830,0.083546
4,0.129088,0.143763,0.101267,0.098094,0.095252,0.125343,0.117536,0.128992,0.060665


In [240]:
xgb_preds = pd.read_csv('../data/xgb_proba.csv')
xgb_preds.head()

,another_bug_proba,break_bug_proba,electro_bug_proba,engine_check_proba,engine_fuel_proba,engine_ignition_proba,engine_overheat_proba,gear_stick_proba,wheel_shake_proba
0,0.104721,0.112521,0.137060,0.090919,0.122933,0.105311,0.106699,0.110659,0.109176
1,0.188389,0.116850,0.096975,0.120398,0.098549,0.088542,0.109467,0.102343,0.078488
2,0.115143,0.113272,0.105387,0.111472,0.097676,0.122911,0.114653,0.112460,0.107026
3,0.109527,0.112851,0.109656,0.136960,0.091574,0.120251,0.090839,0.116508,0.111834
4,0.121198,0.110750,0.130662,0.098771,0.132158,0.112390,0.102490,0.101101,0.090482


In [241]:
final_preds = sum([cb_preds, lgbm_preds , xgb_preds])
final_preds.head()

,another_bug_proba,break_bug_proba,electro_bug_proba,engine_check_proba,engine_fuel_proba,engine_ignition_proba,engine_overheat_proba,gear_stick_proba,wheel_shake_proba
0,0.337219,0.344227,0.353973,0.314100,0.346557,0.322595,0.336272,0.352935,0.292123
1,0.479751,0.442886,0.292707,0.333411,0.280346,0.284728,0.328361,0.325581,0.232230
2,0.348240,0.342305,0.317633,0.329424,0.307271,0.362316,0.348468,0.359000,0.285343
3,0.339210,0.347235,0.317740,0.374427,0.301501,0.341304,0.311298,0.377495,0.289790
4,0.359307,0.372359,0.335765,0.314143,0.331229,0.346983,0.335586,0.364043,0.240584


In [242]:
final_preds['soft_preds'] = final_preds.idxmax(axis=1)
result = final_preds[['soft_preds']].map(lambda x: x.split('_proba')[0])
result

,soft_preds
0,electro_bug
1,another_bug
2,engine_ignition
3,gear_stick
4,break_bug
...,...
580,another_bug
581,gear_stick
582,break_bug
583,gear_stick


In [243]:
result.head(10)

,soft_preds
0,electro_bug
1,another_bug
2,engine_ignition
3,gear_stick
4,break_bug
5,engine_overheat
6,engine_fuel
7,break_bug
8,engine_fuel
9,gear_stick


Ваша задача  — сблендить предсказания этих трех моделей при помощи методики мягкого голосования (soft voting). В качестве ответа отправьте файл, в котором для каждой строчки будет предсказанный ответ, записанный в колонку soft_preds. Названия ответа можно брать из названия соответствующей колонки, обрезая _probа. Порядок строк в высылаемом решении должен быть таким же, как и в файлах с предсказаниями моделей.



In [244]:
result.to_csv('../subs/cars_soft_voting.csv', index=False)
pd.read_csv('../subs/cars_soft_voting.csv').head(10)

,soft_preds
0,electro_bug
1,another_bug
2,engine_ignition
3,gear_stick
4,break_bug
5,engine_overheat
6,engine_fuel
7,break_bug
8,engine_fuel
9,gear_stick


# Задача на 10-ом шаге урока

**Напишите свой Rank Averaging**

Реализуйте процедуру нормализации вероятностей, как описывалось в предыдущем шаге (https://stepik.org/lesson/825510/step/9?unit=829017). Для создания тестового датасета можно воспользоваться кодом:

In [245]:
import numpy as np
import pandas as pd
np.random.seed(42)
proba = np.random.random(100)
df = pd.DataFrame({'id': np.arange(len(proba)), 'proba': proba})
df

,id,proba
0,0,0.374540
1,1,0.950714
2,2,0.731994
3,3,0.598658
4,4,0.156019
...,...,...
95,95,0.493796
96,96,0.522733
97,97,0.427541
98,98,0.025419


Вам нужно посчитать отнормированные предсказания по инструкции из предыдущего степа. Финальный ответ запишите в столбец normalized_proba

In [246]:
import numpy as np
import pandas as pd
df['normalized_proba'] = #your code

SyntaxError: invalid syntax (3750437769.py, line 3)

# Задача на 11-ом шаге урока

**Rank Averaging для 3-х моделей**

В этот раз вам даны вероятности сразу для трех моделей -  proba1, proba2, proba3. Для создания тестового датасета можно использовать код:




In [ ]:
import numpy as np
import pandas as pd
np.random.seed(42)
df = pd.DataFrame({'id': np.arange(100),
                   'proba1': np.random.random(100),
                   'proba2': np.random.random(100),
                   'proba3': np.random.random(100)})

In [ ]:
df

,id,proba1,proba2,proba3
0,0,0.374540,0.031429,0.642032
1,1,0.950714,0.636410,0.084140
2,2,0.731994,0.314356,0.161629
3,3,0.598658,0.508571,0.898554
4,4,0.156019,0.907566,0.606429
...,...,...,...,...
95,95,0.493796,0.349210,0.522243
96,96,0.522733,0.725956,0.769994
97,97,0.427541,0.897110,0.215821
98,98,0.025419,0.887086,0.622890


Посчитайте для каждой модели ее ранг (как в прошлом степе) и потом запишите среднее значение этих трех рангов в колонку mean_rank.

In [ ]:
import numpy as np
import pandas as pd

df['mean_rank'] = #your code